ENH: #59 - annotation form

lightandluck opened the pull request:
~**DO NOT MERGE UNTIL #82 IS MERGED**~

This PR is branched from everyone's work on #82 which we agreed would make more sense to branch from rather than `master`. It addresses #59. **It is merely cosmetic.** I created buttons for the annotation form, cleaned up some of the layout, added a notes field and tooltips.

~**Updating annotations with notes doesn't work yet.** Most changes were done only in `annotation-form.tsx` and `styles.css`. I added some class names on components outside of `annotation-form.tsx` to help with styling layout but didn't touch anything else to avoid conflicts.~

[Demo](http://edgi-web-monitor-ui.herokuapp.com/page/9da29641-fd4e-4767-bf3f-f808967f7b15)

Mr0grog commented:
Rebased now that #82 is merged.

Mr0grog commented:
There is probably a little more cleanup that could be done here before merging.

I’d also like to look into alternatives to [react-lightweight-tooltip](https://github.com/mcumpl/react-lightweight-tooltip). It is triggering deprecation warnings from React and looks like it hasn’t been updated in almost a year (there are 183 open PRs! Most are from Greenkeeper, but not all are). Normally I wouldn’t be worried about non-recent updates from a simple library like this, but the underlying framework—React—changes fast and in big ways. Those warnings will soon become errors. That giant pile of simple-but-unaddressed PRs is also a clear signal that the library is unmaintained and probably not taking patches.

Mr0grog commented:
Merged TS → JS conversion work in (rebasing would surely have been nuts).

lightandluck commented:
> I’d also like to look into alternatives to react-lightweight-tooltip

Absolutely. I ended up using that component because it was the 3rd or 4th thing I tried that finally worked with Typescript out of the box. There were huge component libraries out there, but I didn't want to figure out how to create the typings for them. But now that TS is not in the project anymore, all of that is moot. 

I looked through the Data Together repo to try to find a tooltip component to reuse, but there didn't seem to be one there. I found [rc-tooltip](https://github.com/react-component/tooltip) that looks better maintained, but I couldn't get working before because of TS. I can switch to that for this PR.

On the subject, would it make sense to use the larger component libraries? These all look very active: [React-Bootstrap](https://github.com/react-bootstrap), [Material UI](https://github.com/callemall/material-ui), [Pivotal UI](https://github.com/pivotal-cf/pivotal-ui). I've thrown out the idea of using tabs for workflow in future iterations. So it might make sense to use a library that we can just pull different UI components from. Pivotal seems ideal, because you don't have to import every component, but can import individual ones and a small core library to run it.

I don't have any experience using any of them so if anyone does or has better suggestions, it would be appreciated. @b5, you're the only person I know of in EDGI that could weigh in here with the most experience. Would be grateful for your input.

lightandluck commented:
@Mr0grog 
Will clean up and go through and address all the FIXME comments you left.

lightandluck commented:
Ready for review

Mr0grog commented:
@lightandluck going to try and review this later tonight, but can you rebase on or merge master and resolve the conflicts? Thanks.

lightandluck commented:
I [did](https://github.com/edgi-govdata-archiving/web-monitoring-ui/pull/92/commits/455888fceb5dbeb6e3fb0f8f53215d7ab3cfdd0a) after you merged #94 and the hotfix. I don't see anything else. Is there something I'm missing?

lightandluck commented:
Some quick points to describe changes I made:
- I turned the header into a link, so we wouldn't have to look at crazy looking urls and reclaim some space
- got rid of extraneous margins, so that things lined up better
- got rid of margins around the body and removed padding from the header to reclaim more space
- I used this [react-tooltip](https://github.com/wwayne/react-tooltip). Seems a little better maintained. Also, only thing I could get working correctly. I think it works a lot better than the previous component I used
- I moved the version selector controls to be on top of the diff view. It made more sense to me to have the controls near the view. I also changed the ordering of the controls to mimic versionista. Just trying to keep things that are associated in proximity of each other.
- I left a small margin between the header and body, leaving a small green border. Graphically, it just gives the eye somewhere to rest, otherwise we're staring at a blinding white screen with text
- Added back in icons to break up all the text
- Thanks for getting the notes field working!

Mr0grog commented:
I'll take a look, then. GitHub is showing me a cannot-auto-merge-due-to-conflicts message.

lightandluck commented:
Am I looking at the right screen?
![screen shot 2017-08-09 at 6 34 15 pm](https://user-images.githubusercontent.com/3014813/29150746-6fb04440-7d31-11e7-857f-d000e54a6b22.png)

Mr0grog commented:
Hahaha, we are getting conflicting messages about conflicts :P

Mr0grog commented on src/css/styles.css:
```diff
@@ -79,57 +93,227 @@ body {
+
+.annotation-form--collapsed .__react_component_tooltip {
+    display: block!important;
+}
```
It seems like we’d be better off to add a `--expanded` class and use that to create just one rule instead of two. Alternatively, you could also just not render the`<Tooltip>` element over on the JS side if the form is expanded.

Mr0grog replied:
Oh, it looks like react-tooltip has a `disable` prop, so you could just do this in the `render()` function and not worry about clobbering stuff with CSS:

```jsx
<Tooltip disable={!this.props.collapsed} />
```

lightandluck replied:
Done! Really cool to use the prop, instead of css.

lightandluck commented on src/css/styles.css:
```diff
@@ -1,25 +1,27 @@
-
-.lnk-action {
-    margin-right: 17px;
+    padding: 1em;
```
I changed the padding to be 15px, so we wouldn't have to add another rule to fight Bootstrap. Less to get rid of when we migrate away as well.

lightandluck commented on src/css/styles.css:
```diff
@@ -79,57 +93,227 @@ body {
-    margin-right: 25px;
+/* Toggle buttons */
+.toggle {
+    display: none;
```
I changed it to use `opacity: 0; width: 0` instead.

lightandluck commented on src/css/styles.css:
```diff
@@ -79,57 +93,227 @@ body {
+    position: relative;
+    user-select: none;
+    transition: all .1s;
+    width: 3.5em;
```
Went with `flex-shrink: 0`. Much nicer

lightandluck commented on src/components/annotation-form.jsx:
```diff
@@ -17,79 +18,101 @@ import {Version} from '../services/web-monitoring-db';
+    return (
+      <form className={classes.join(' ')}>
+        <div className="annotation-form__signifiers">
+          <div className="signifier-container">
```
Done

lightandluck commented on src/components/annotation-form.jsx:
```diff
@@ -17,79 +18,101 @@ import {Version} from '../services/web-monitoring-db';
-        );
-    }
+        <textarea
+          name="further-notes"
```
Done

lightandluck commented on src/components/annotation-form.jsx:
```diff
@@ -17,79 +18,101 @@ import {Version} from '../services/web-monitoring-db';
+    <li className="signifier-list-item">
+      <input type="checkbox" className="toggle" id={name} checked={checked} onChange={changeHandler} />
+      <label data-tip={children} data-for={`${name}-tip`} className="toggle-btn" data-tg-on={fieldNumber} htmlFor={name} />
+      <span className="info-text">{children}</span>
```
I was able to insert the span into the label so that it would all be clickable. Also, realized I could show or hide it using conditional rendering rather than css. Starting to get it, React rocks!

Still had to keep the `input` and `label` separate because of the unique styling. The links you recommended all do the same as well. I added a unique ID for the fields

lightandluck replied:
I did remove a bunch of unnecessary crud too

lightandluck commented on src/components/annotation-form.jsx:
```diff
@@ -17,79 +18,101 @@ import {Version} from '../services/web-monitoring-db';
+        <div className="annotation-form__signifiers">
+          <div className="signifier-container">
+            <h5>Individual Page Changes</h5>
+            <ul className="signifier-list">
```
I'm using a list because the `li` elements are helping to contain and style the checkbox controls. See comment below about why they're kept separate.

Did change it to `<ol>`

lightandluck commented on src/components/annotation-form.jsx:
```diff
@@ -17,79 +18,101 @@ import {Version} from '../services/web-monitoring-db';
+      <input type="checkbox" className="toggle" id={name} checked={checked} onChange={changeHandler} />
+      <label data-tip={children} data-for={`${name}-tip`} className="toggle-btn" data-tg-on={fieldNumber} htmlFor={name} />
+      <span className="info-text">{children}</span>
+      <Tooltip id={`${name}-tip`} place="top" />
```
Yup, was able to pull `<Tooltip>` out and change the effect

lightandluck commented on src/components/change-view.jsx:
```diff
@@ -111,9 +111,9 @@ export default class ChangeView extends React.Component {
         return (
-            <div className="change-view">
-                {this.renderVersionSelector(page)}
+            <div className="change-view-container">
```
This was an artifact from way before when I initially branched. I changed it back to reflect convention and also got rid of unnecessary styles associated with it.

lightandluck commented on src/components/page-details.jsx:
```diff
@@ -81,8 +81,7 @@ export default class PageDetails extends React.Component {
                     <div className="col-md-9">
-                        <h2 className="diff-title">{page.title}</h2>
-                        <a className="diff_page_url" href={page.url} target="_blank" rel="noopener">{page.url}</a>
+                        <h2 className="page-title"><a className="diff_page_url" href={page.url} target="_blank" rel="noopener">{page.title}</a></h2>
```
Yes, not a problem.

Mr0grog replied:
…?

lightandluck replied:
Sorry, I was figuring out my formatting settings and had to restore some files. Sound be good now! lol

lightandluck commented on src/css/diff.css:
```diff
@@ -10,8 +10,14 @@ ins {
+    width: 100%;
+}
+
+.side-by-side-render > iframe:first-child {
```
Haven't heard of `:first-of-type`. Makes more sense here and made the change.

lightandluck commented:
> I do kinda wonder if it’s no longer clear that clicking it is going to take you to the actual page.

I'm not as concerned because we can just tell the analysts as they are being trained. But, I agree we should discuss with analysts. It really does help tighten up the layout, I think.

> I think we have an information hierarchy problem here

Yea, you're right about the hierarchy. I wasn't thinking about version selection changing the annotations. The way I was looking at it is how they're currently using versionista, where they don't change the versions as much, but only look at diffs. Our system is much more powerful and allows them to navigate through all the history in one place, but I don't think that will be used much in the first pass. But definitely will be used in second pass more for sure.

I'm going to leave it as is for now. Also, your timeline (which should definitely be used) will change all this anyway. As I mentioned, I think after this is merged, it would be a good time to get feedback from lead analysts. We can hammer out details then.

One extra change I did was switching the pagination to use flexbox as well. Just a heads up.

Ready for review!

b5 commented:
Ehhhyo,

Nice work finding dem libraries @lightandluck!

Looks like material-ui uses typescript, and is structured with the intent that you won't wander far outside of it's component library. I'd knock that one off the list straight out the gates. Also, styles are inlined into the js, which are subject of rigorous debate. 

Pivotal-UI & React-Bootstrap both look like great options. After browsing through their components, they're both great react projects IMHO. I'd say drawing from either would be lovely, I'd go with bootstrap if only because I find it less ugly. 

One suggestion, moving to webpack 2 as the transpiler will give you [tree-shaking](https://webpack.js.org/guides/tree-shaking/), which _really_ shines when you start pulling in these bloated UI libraries. It'll mean that you can import with impunity, knowing that the production build will eliminate dead code for you.

In another neck of the woods, I've opened up datatogether/context#44 to see if we can't make some fun progress of our own as well ;)

Mr0grog commented on src/css/styles.css:
```diff
@@ -79,57 +101,186 @@ body {
+.toggle {
+    opacity: 0;
+    width: 0;
+}
```
This is definitely an improvement in that you can tab to the checkbox, but you still can’t actually see what checkbox you’re focused on. You might try adding a rule like:

```css
.toggle:focus + .toggle-btn::before {
  outline: 2px solid aliceblue;               // Generic case, whatever you want
  outline: -moz-mac-focusring auto 3px;       // Firefox on OS X (no Windows-specific equivalent, ¯\_(ツ)_/¯)
  outline: -webkit-focus-ring-color auto 5px; // WebKit + Blink
}
```

Obviously you can do whatever focus-specific styling you want; the above just gets you something close-ish to what the system does by default.

lightandluck replied:
Ah, I wasn't thinking about tabbing through them at all. Fixed

Mr0grog commented on src/components/annotation-form.jsx:
```diff
@@ -17,79 +18,122 @@ import {Version} from '../services/web-monitoring-db';
+        data-tip={children}
+        data-fieldNumber={fieldNumber}
+        htmlFor={uuid}>
+          { collapsed ? "" : <span>{children}</span> }
```
This is fine and there’s no need to change it back, but if you are working in more performance-sensitive areas in the future, it’s *dramatically* faster and more efficient to alter this with CSS than by adding or removing elements. (If you look at good scrolling list implementations or things like CodeMirror, for example, you’ll see they reuse elements that have moved off-screen and put new text in them.)

lightandluck replied:
Arg, I'm learning just enough to be dangerous 😟 , duly noted

lightandluck commented:
Fixed up dem changes

Mr0grog commented:
Going to merge this. @b5: #87 or #88 is probably a better place for the UI libraries discussion.
